contracts/oracles/yieldspace/PoolOracle.sol

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.6;
+
+import {IPool} from "@yield-protocol/yieldspace-interfaces/IPool.sol";
+import {IPoolOracle} from "./IPoolOracle.sol";
+
+/**
+ * @title PoolOracle
+ * @author Bruno Bonanno
+ * @dev This contract collects data from different YieldSpace pools to compute a TWAR using a SMA (https://www.investopedia.com/terms/s/sma.asp)
+ * Adapted from https://github.com/Uniswap/v2-periphery/blob/master/contracts/examples/ExampleSlidingWindowOracle.sol
+ */
+//solhint-disable not-rely-on-time
+contract PoolOracle is IPoolOracle {
+    event ObservationRecorded(address indexed pool, uint256 index, Observation observation);
+
+    error NoObservationsForPool(address pool);
+    error MissingHistoricalObservation(address pool);
+    error InsufficientElapsedTime(address pool, uint256 elapsedTime);
+
+    struct Observation {
+        uint256 timestamp;
+        uint256 ratioCumulative;
+    }
+
+    // the desired amount of time over which the moving average should be computed, e.g. 24 hours
+    uint256 public immutable windowSize;
+    // the number of observations stored for each pool, i.e. how many ratio observations are stored for the window.
+    // as granularity increases from 1, more frequent updates are needed, but moving averages become more precise.
+    // averages are computed over intervals with sizes in the range:
+    //   [windowSize - (windowSize / granularity) * 2, windowSize]
+    // e.g. if the window size is 24 hours, and the granularity is 24, the oracle will return the TWAR for
+    //   the period:
+    //   [now - [22 hours, 24 hours], now]
+    uint256 public immutable granularity;
+    // this is redundant with granularity and windowSize, but stored for gas savings & informational purposes.
+    uint256 public immutable periodSize;
+    // this is to avoid using values that are too close in time to the current observation
+    uint256 public immutable minTimeElapsed;
+
+    // mapping from pool address to a list of ratio observations of that pool
+    mapping(address => Observation[]) public poolObservations;
+
+    constructor(
+        uint256 windowSize_,
+        uint256 granularity_,
+        uint256 minTimeElapsed_
+    ) {
+        require(granularity_ > 1, "GRANULARITY");
+        require((periodSize = windowSize_ / granularity_) * granularity_ == windowSize_, "WINDOW_NOT_EVENLY_DIVISIBLE");
+        windowSize = windowSize_;
+        granularity = granularity_;
+        minTimeElapsed = minTimeElapsed_;
+    }
+
+    // returns the index of the observation corresponding to the given timestamp
```

Comment:
Please use full natspec for functions, so that docs can be automatically generated.